File: scripts/differential/java/P2T3R.java
package app.freerouting.datastructures;

import app.freerouting.geometry.planar.FortyfiveDegreeBoundingDirections;
import app.freerouting.geometry.planar.IntBox;
import app.freerouting.geometry.planar.IntOctagon;
import app.freerouting.geometry.planar.OrthogonalBoundingDirections;
import app.freerouting.geometry.planar.RegularTileShape;
import app.freerouting.geometry.planar.ShapeBoundingDirections;
import app.freerouting.geometry.planar.TileShape;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Randomised Java ground truth for MinAreaTree/ShapeTree. Twin: `p2t3r`.
 *
 * <p>args: ops seed mode(0=orthogonal,1=45-degree) dumpEvery insertPct
 *
 * <p>Drives `insert(Storable)` (so the tree itself applies boundingDirections),
 * `remove(Leaf[])` with deliberate `null` holes, in-place re-keying of live leaves the way
 * `ShapeSearchTree` does, `overlaps`, `toArray` and `distanceToRoot`. Every random draw is
 * mirrored draw-for-draw by `rust/src/bin/p2t3r.rs`.
 */
public class P2T3R {

  static long state;

  static long next() {
    state ^= state << 13;
    state ^= state >>> 7;
    state ^= state << 17;
    return state;
  }

  static int rnd(int bound) {
    return (int) Long.remainderUnsigned(next(), bound);
  }

  static class Obj implements ShapeTree.Storable {
    final int id;
    final TileShape[] shapes;
    ShapeTree.Leaf[] entries;

    Obj(int id, TileShape... shapes) {
      this.id = id;
      this.shapes = shapes;
    }

    public int treeShapeCount(ShapeTree tree) {
      return shapes.length;
    }

    public TileShape getTreeShape(ShapeTree tree, int index) {
      return shapes[index];
    }

    public void setSearchTreeEntries(ShapeTree.Leaf[] entries, ShapeTree tree) {
      this.entries = entries;
    }

    public int compareTo(Object other) {
      return Integer.compare(this.id, ((Obj) other).id);
    }

    public String toString() {
      return "#" + id;
    }
  }

  static String boundsStr(RegularTileShape s) {
    if (s instanceof IntBox b) {
      return "B(" + b.ll.x + "," + b.ll.y + "," + b.ur.x + "," + b.ur.y + ")";
    }
    IntOctagon o = (IntOctagon) s;
    return "O("
        + o.leftX + "," + o.bottomY + "," + o.rightX + "," + o.topY + ","
        + o.upperLeftDiagonalX + "," + o.lowerRightDiagonalX + ","
        + o.lowerLeftDiagonalX + "," + o.upperRightDiagonalX + ")";
  }

  static void dump(ShapeTree.TreeNode node, String prefix, StringBuilder sb) {
    if (node instanceof ShapeTree.Leaf leaf) {
      sb.append(prefix).append("L ").append(leaf.object).append("/")
          .append(leaf.shapeIndexInObject).append(" ").append(boundsStr(leaf.boundingShape))
          .append("\n");
    } else {
      ShapeTree.InnerNode inner = (ShapeTree.InnerNode) node;
      sb.append(prefix).append("I ").append(boundsStr(inner.boundingShape)).append("\n");
      dump(inner.firstChild, prefix + " ", sb);
      dump(inner.secondChild, prefix + " ", sb);
    }
  }

  static void printTree(MinAreaTree tree, String label) {
    System.out.println("== " + label + " n=" + tree.size() + " rootNull=" + (tree.root == null));
    if (tree.root == null) {
      return;
    }
    StringBuilder sb = new StringBuilder();
    dump(tree.root, "", sb);
    System.out.print(sb);
  }

  static String key(ShapeTree.Leaf leaf) {
    return leaf.object + "/" + leaf.shapeIndexInObject;
  }

  public static void main(String[] args) {
    int ops = args.length > 0 ? Integer.parseInt(args[0]) : 200;
    long seed = args.length > 1 ? Long.parseLong(args[1]) : 1L;
    int mode = args.length > 2 ? Integer.parseInt(args[2]) : 0;
    int dumpEvery = args.length > 3 ? Integer.parseInt(args[3]) : 1;
    int insertPct = args.length > 4 ? Integer.parseInt(args[4]) : 62;
    state = seed == 0 ? 0x9E3779B97F4A7C15L : seed;

    ShapeBoundingDirections dirs =
        mode == 0
            ? OrthogonalBoundingDirections.INSTANCE
            : FortyfiveDegreeBoundingDirections.INSTANCE;
    MinAreaTree tree = new MinAreaTree(dirs);

    List<ShapeTree.Leaf> live = new ArrayList<>();
    int nextId = 1;

    for (int step = 0; step < ops; step++) {
      int roll = rnd(100);
      if (live.isEmpty() || roll < insertPct) {
        // `ShapeTree.insert(Storable)`: the TREE applies boundingDirections to each tree shape.
        int shapeCount = 1 + rnd(3);
        TileShape[] shapes = new TileShape[shapeCount];
        for (int k = 0; k < shapeCount; k++) {
          int x = rnd(2000) - 500;
          int y = rnd(2000) - 500;
          int w = 1 + rnd(150);
          int h = 1 + rnd(150);
          shapes[k] = new IntBox(x, y, x + w, y + h);
        }
        Obj o = new Obj(nextId++, shapes);
        tree.insert(o);
        StringBuilder ib = new StringBuilder("op" + step + " insert " + o + " x" + shapeCount);
        for (ShapeTree.Leaf lf : o.entries) {
          if (lf == null) {
            ib.append(" -");
          } else {
            live.add(lf);
            ib.append(" ").append(key(lf));
          }
        }
        System.out.println(ib);
      } else if (roll < insertPct + 12) {
        // In-place re-key of a live leaf: `ShapeSearchTree.java:150,214-215,221,294-295,325-326,
        // 342-343` assign leaf.object / leaf.shapeIndexInObject without touching bounds or links.
        int idx = rnd(live.size());
        int newIndex = rnd(4);
        ShapeTree.Leaf lf = live.get(idx);
        Obj newOwner = new Obj(nextId++);
        System.out.println(
            "op" + step + " rekey " + key(lf) + " -> " + newOwner + "/" + newIndex);
        lf.object = newOwner;
        lf.shapeIndexInObject = newIndex;
      } else {
        // `remove(Leaf[])` on an array with `null` holes, as `reuseEntriesAfterCutout` leaves
        // behind (ShapeSearchTree.java:327,344) and `SearchTreeManager.remove` passes on.
        int batch = 1 + rnd(4);
        ShapeTree.Leaf[] batchArr = new ShapeTree.Leaf[batch];
        StringBuilder rb = new StringBuilder("op" + step + " remove");
        for (int k = 0; k < batch; k++) {
          boolean hole = rnd(5) == 0;
          if (hole || live.isEmpty()) {
            batchArr[k] = null;
            rb.append(" -");
          } else {
            int idx = rnd(live.size());
            ShapeTree.Leaf lf = live.remove(idx);
            batchArr[k] = lf;
            rb.append(" ").append(key(lf));
          }
        }
        System.out.println(rb);
        tree.remove(batchArr);
      }

      if (step % dumpEvery == 0 || step == ops - 1) {
        printTree(tree, "op" + step);
      }

      for (int q = 0; q < 2; q++) {
        int x = rnd(2000) - 500;
        int y = rnd(2000) - 500;
        int w = 1 + rnd(600);
        int h = 1 + rnd(600);
        RegularTileShape query = new IntBox(x, y, x + w, y + h).boundingShape(dirs);
        Set<ShapeTree.Leaf> res = tree.overlaps(query);
        StringBuilder sb = new StringBuilder("q" + q + " " + boundsStr(query) + " ->");
        for (ShapeTree.Leaf l : res) {
          sb.append(" ").append(key(l));
        }
        System.out.println(sb);
      }

      ShapeTree.Leaf[] arr = tree.toArray();
      StringBuilder ta = new StringBuilder("toArray");
      for (ShapeTree.Leaf l : arr) {
        ta.append(" ").append(key(l));
      }
      System.out.println(ta);

      // distanceToRoot for every live leaf; -1 stands in for the one-element NPE case.
      StringBuilder dr = new StringBuilder("depths");
      for (ShapeTree.Leaf l : arr) {
        dr.append(" ").append(tree.size() == 1 ? -1 : l.distanceToRoot());
      }
      System.out.println(dr);
    }
  }
}

File: scripts/differential/rust/src/bin/p2t3r.rs
//! Rust twin of `../../java/P2T3R.java`: randomised differential driver for
//! `fr_board::datastructures::ShapeTree` against Java's `ShapeTree`/`MinAreaTree`.
//!
//! args: ops seed mode(0=orthogonal,1=45-degree) dumpEvery insertPct
//!
//! Every random draw is mirrored draw-for-draw by the Java side; the two programs must print
//! byte-identical stdout.

use fr_board::datastructures::{LeafId, Node, NodeId, ShapeTree};
use fr_board::ids::{ItemId, TreeObject};
use fr_geometry::bounding_directions::ShapeBoundingDirections;
use fr_geometry::int_box::IntBox;
use fr_geometry::regular_tile_shape::RegularTileShape;
use fr_geometry::tile_shape::TileShape;

struct Rng(u64);

impl Rng {
    fn next(&mut self) -> u64 {
        self.0 ^= self.0 << 13;
        self.0 ^= self.0 >> 7;
        self.0 ^= self.0 << 17;
        self.0
    }

    /// Java `(int) Long.remainderUnsigned(next(), bound)`.
    fn rnd(&mut self, bound: u64) -> i64 {
        (self.next() % bound) as i64
    }
}

fn bounds_str(s: &RegularTileShape) -> String {
    match s {
        RegularTileShape::Box(b) => format!("B({},{},{},{})", b.ll.x, b.ll.y, b.ur.x, b.ur.y),
        RegularTileShape::Octagon(o) => format!(
            "O({},{},{},{},{},{},{},{})",
            o.left_x,
            o.bottom_y,
            o.right_x,
            o.top_y,
            o.upper_left_diagonal_x,
            o.lower_right_diagonal_x,
            o.lower_left_diagonal_x,
            o.upper_right_diagonal_x
        ),
    }
}

fn obj_str(o: &TreeObject) -> String {
    match o {
        TreeObject::Item(ItemId(n)) => format!("#{n}"),
        other => format!("{other:?}"),
    }
}

fn key(tree: &ShapeTree<TreeObject>, leaf: LeafId) -> String {
    let e = tree.leaf_entry(leaf);
    format!("{}/{}", obj_str(&e.object), e.shape_index)
}

fn dump(tree: &ShapeTree<TreeObject>, id: NodeId, prefix: &str, out: &mut String) {
    match tree.node(id) {
        Node::Leaf {
            bounds,
            object,
            shape_index,
            ..
        } => {
            out.push_str(&format!(
                "{prefix}L {}/{} {}\n",
                obj_str(object),
                shape_index,
                bounds_str(bounds)
            ));
        }
        Node::Inner {
            bounds,
            first_child,
            second_child,
            ..
        } => {
            out.push_str(&format!("{prefix}I {}\n", bounds_str(bounds)));
            let deeper = format!("{prefix} ");
            dump(tree, *first_child, &deeper, out);
            dump(tree, *second_child, &deeper, out);
        }
        Node::Free { .. } => unreachable!("a reachable node is never a free slot"),
    }
}

fn print_tree(tree: &ShapeTree<TreeObject>, label: &str) {
    println!(
        "== {label} n={} rootNull={}",
        tree.leaf_count(),
        tree.root().is_none()
    );
    if let Some(root) = tree.root() {
        let mut s = String::new();
        dump(tree, root, "", &mut s);
        print!("{s}");
    }
}

fn main() {
    let a: Vec<String> = std::env::args().skip(1).collect();
    let ops: usize = a.first().map(|s| s.parse().unwrap()).unwrap_or(200);
    let seed: i64 = a.get(1).map(|s| s.parse().unwrap()).unwrap_or(1);
    let mode: i32 = a.get(2).map(|s| s.parse().unwrap()).unwrap_or(0);
    let dump_every: usize = a.get(3).map(|s| s.parse().unwrap()).unwrap_or(1);
    let insert_pct: i64 = a.get(4).map(|s| s.parse().unwrap()).unwrap_or(62);
    let mut rng = Rng(if seed == 0 {
        0x9E37_79B9_7F4A_7C15
    } else {
        seed as u64
    });

    let dirs = if mode == 0 {
        ShapeBoundingDirections::Orthogonal
    } else {
        ShapeBoundingDirections::FortyfiveDegree
    };
    let mut tree: ShapeTree<TreeObject> = ShapeTree::new(dirs);
    let mut live: Vec<LeafId> = Vec::new();
    let mut next_id: u32 = 1;

    for step in 0..ops {
        let roll = rng.rnd(100);
        if live.is_empty() || roll < insert_pct {
            // `ShapeTree::insert_tiles` — the port of `ShapeTree.insert(Storable)`: the TREE
            // applies its bounding directions to each tile shape.
            let shape_count = 1 + rng.rnd(3) as usize;
            let mut shapes: Vec<TileShape> = Vec::new();
            for _ in 0..shape_count {
                let x = rng.rnd(2000) - 500;
                let y = rng.rnd(2000) - 500;
                let w = 1 + rng.rnd(150);
                let h = 1 + rng.rnd(150);
                shapes.push(TileShape::Box(IntBox::from_coords(
                    x as i32,
                    y as i32,
                    (x + w) as i32,
                    (y + h) as i32,
                )));
            }
            let o = TreeObject::Item(ItemId(next_id));
            next_id += 1;
            let entries = tree.insert_tiles(o, &shapes);
            let mut ib = format!("op{step} insert {} x{shape_count}", obj_str(&o));
            for entry in entries {
                match entry {
                    None => ib.push_str(" -"),
                    Some(id) => {
                        live.push(id);
                        ib.push_str(&format!(" {}", key(&tree, id)));
                    }
                }
            }
            println!("{ib}");
        } else if roll < insert_pct + 12 {
            // In-place re-key of a live leaf: `ShapeSearchTree.java:150,214-215,221,294-295,
            // 325-326,342-343`.
            let idx = rng.rnd(live.len() as u64) as usize;
            let new_index = rng.rnd(4) as usize;
            let leaf = live[idx];
            let new_owner = TreeObject::Item(ItemId(next_id));
            next_id += 1;
            println!(
                "op{step} rekey {} -> {}/{new_index}",
                key(&tree, leaf),
                obj_str(&new_owner)
            );
            tree.set_leaf_entry(leaf, new_owner, new_index);
        } else {
            // `remove_opt` — the port of `remove(Leaf[])` on an array with `null` holes.
            let batch = 1 + rng.rnd(4) as usize;
            let mut batch_arr: Vec<Option<LeafId>> = Vec::new();
            let mut rb = format!("op{step} remove");
            for _ in 0..batch {
                let hole = rng.rnd(5) == 0;
                if hole || live.is_empty() {
                    batch_arr.push(None);
                    rb.push_str(" -");
                } else {
                    let idx = rng.rnd(live.len() as u64) as usize;
                    let leaf = live.remove(idx);
                    rb.push_str(&format!(" {}", key(&tree, leaf)));
                    batch_arr.push(Some(leaf));
                }
            }
            println!("{rb}");
            tree.remove_opt(&batch_arr);
        }

        if step % dump_every == 0 || step == ops - 1 {
            print_tree(&tree, &format!("op{step}"));
        }

        for q in 0..2 {
            let x = rng.rnd(2000) - 500;
            let y = rng.rnd(2000) - 500;
            let w = 1 + rng.rnd(600);
            let h = 1 + rng.rnd(600);
            let tile = TileShape::Box(IntBox::from_coords(
                x as i32,
                y as i32,
                (x + w) as i32,
                (y + h) as i32,
            ));
            let query = tree
                .bounding_shape(&tile)
                .expect("a box is always boundable");
            let res = tree.overlaps(&query);
            let mut s = format!("q{q} {} ->", bounds_str(&query));
            for e in &res {
                s.push_str(&format!(" {}/{}", obj_str(&e.object), e.shape_index));
            }
            println!("{s}");
        }

        let arr = tree.to_array();
        let mut ta = String::from("toArray");
        for l in &arr {
            ta.push_str(&format!(" {}", key(&tree, *l)));
        }
        println!("{ta}");

        let mut dr = String::from("depths");
        for l in &arr {
            let d: i64 = if tree.leaf_count() == 1 {
                -1
            } else {
                tree.distance_to_root(*l) as i64
            };
            dr.push_str(&format!(" {d}"));
        }
        println!("{dr}");
    }
}
